tools/run_tests/xds_k8s_test_driver/tests/authz_test.py

```diff
@@ -0,0 +1,313 @@
+                        "regexMatch": "never-match-host",
+                    },
+                },
+            },
+            {
+                "destinations": {
+                    "hosts": [f"*:{self.server_xds_port}"],
+                    "ports": [1],
+                    "httpHeaderMatch": {
+                        "headerName": "test",
+                        "regexMatch": "never-match-port",
+                    },
+                },
+            },
+            {
+                "sources": {
+                    "principals": ["*"],
+                },
+                "destinations": {
+                    "hosts": [f"*:{self.server_xds_port}"],
+                    "ports": [self.server_port],
+                    "httpHeaderMatch": {
+                        "headerName": "test",
+                        "regexMatch": "principal-present",
+                    },
+                },
+            },
+            {
+                "sources": [{
+                    "principals": [
+                        f"spiffe://{self.project}.svc.id.goog/not/the/client",
+                    ],
+                }, {
+                    "principals": [
+                        f"spiffe://{self.project}.svc.id.goog/not/the/client",
+                        f"spiffe://{self.project}.svc.id.goog/ns/"
+                        f"{self.client_namespace}/sa/{self.client_name}",
+                    ],
+                }],
+                "destinations": {
+                    "hosts": [f"*:{self.server_xds_port}"],
+                    "ports": [self.server_port],
+                    "httpHeaderMatch": {
+                        "headerName": "test",
+                        "regexMatch": "match-principal",
+                    },
+                },
+            },
+            {
+                "sources": {
+                    "principals": [
+                        f"spiffe://{self.project}.svc.id.goog/not/the/client",
+                    ],
+                },
+                "destinations": {
+                    "hosts": [f"*:{self.server_xds_port}"],
+                    "ports": [self.server_port],
+                    "httpHeaderMatch": {
+                        "headerName": "test",
+                        "regexMatch": "never-match-principal",
+                    },
+                },
+            },
+        ]
+
+    def configure_and_assert(self, test_client: _XdsTestClient,
+                             test_metadata_val: Optional[str],
+                             status_code: grpc.StatusCode) -> None:
+        # Swap method type every sub-test to avoid mixing results
+        rpc_type = self.next_rpc_type
+        if rpc_type is None:
+            stats = test_client.get_load_balancer_accumulated_stats()
+            for t in self.RPC_TYPE_CYCLE:
+                if not stats.stats_per_method[t].rpcs_started:
+                    rpc_type = t
+            assert rpc_type is not None, "All RPC types already used"
```

Comment:
If we specify a start `self.next_rpc_type` value, maybe we can remove line 165-170? We know what the test_client config is, and we can pin `self.next_rpc_type` to `EMPTY_CALL`, for example.

Ah, this if is only true for the first ever invocation. I thought that `stats.stats_per_method[t].rpcs_started` will be checked for every call to `configure_and_assert`, my bad.

Reply:
We totally can get rid of those lines, but that requires _coordination_ and _alignment_ with code quite far away and if you get it wrong the test is _invalid_. It seemed worth the few lines of code. But I totally hard-coded it initially.

I think you are getting a better understanding of how weird what I'm doing is. But given the current interop clients...